src/python/pants/backend/jvm/tasks/coursier_resolve.py

```diff
@@ -0,0 +1,478 @@
+    super(CoursierMixin, cls).register_options(register)
+
+  @classmethod
+  def subsystem_dependencies(cls):
+    return super(CoursierMixin, cls).subsystem_dependencies() + (JarDependencyManagement, CoursierSubsystem)
+
+  @staticmethod
+  def _compute_jars_to_resolve_and_to_exclude(jars, artifact_set, manager):
+    """
+
+    :param jars_by_key:
+    :param artifact_set:
+    :param manager:
+    :return:
+    """
+
+    if artifact_set is None:
+      artifact_set = PinnedJarArtifactSet()
+
+    jars_by_key = OrderedDict()
+    for jar in jars:
+      jars_for_the_key = jars_by_key.setdefault((jar.org, jar.name), [])
+      jars_for_the_key.append(jar)
+
+    jars_to_resolve = []
+    exclude_args = set()
+
+    untouched_pinned_artifact = set(M2Coordinate.create(x) for x in artifact_set)
+
+    for k, jar_list in jars_by_key.items():
+      for i, dep in enumerate(jar_list):
+        direct_coord = M2Coordinate.create(dep)
+
+        if direct_coord in artifact_set:
+          managed_coord = artifact_set[direct_coord]
+          untouched_pinned_artifact.remove(managed_coord)
+
+          if direct_coord.rev != managed_coord.rev:
+            # It may be necessary to actually change the version number of the jar we want to resolve
+            # here, because overrides do not apply directly (they are exclusively transitive). This is
+            # actually a good thing, because it gives us more control over what happens.
+            coord = manager.resolve_version_conflict(managed_coord, direct_coord, force=dep.force)
+
+            # Once a version is settled, we force it anyway
+            jar_list[i] = dep.copy(rev=coord.rev, force=True)
+
+      jars_to_resolve.extend(jar_list)
+      for jar in jar_list:
+        for ex in jar.excludes:
+          # `--` means exclude. See --soft-exclude-file in `coursier fetch --help`
+          ex_arg = "{}:{}--{}:{}".format(jar.org, jar.name, ex.org, ex.name)
+          exclude_args.add(ex_arg)
+
+    return jars_to_resolve, exclude_args, untouched_pinned_artifact
+
+  @property
+  def cache_target_dirs(self):
+    return True
+
+  def resolve(self, targets, compile_classpath):
+    """
+
+    :param targets: a collection of targets to do 3rdparty resolve against
+    :param compile_classpath: classpath product that holds the resolution result. IMPORTANT: this parameter will be changed.
+    :return:
+    """
+
+    coursier_subsystem_instance = CoursierSubsystem.global_instance()
+
+    coursier_jar = coursier_subsystem_instance.bootstrap_coursier()
+
+    manager = JarDependencyManagement.global_instance()
+
+    jar_targets = manager.targets_by_artifact_set(targets)
+
+    for artifact_set, target_subset in jar_targets.items():
+      # TODO(wisechengyi): this is the only place we are using IvyUtil method, which isn't specific to ivy really.
+      raw_jar_deps, global_excludes = IvyUtils.calculate_classpath(target_subset)
+
+      with self.invalidated(target_subset,
```

Comment:
A best practice for resolve blocks is to keep them as small and clear as possible, with almost all logic delegated to functions. Please break this up as high priority, as it should make this easier to review for correctness.

Reply:
trimmed. let me know if it's still hard to reason.

Reply:
It's still quite large, but much clearer. Thanks.